azurerm/internal/sdk/locations.go

```diff
@@ -0,0 +1,60 @@
+package sdk
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+type SupportedLocations struct {
+	// Locations is a list of Locations which are supported on this Azure Endpoint.
+	// This could be nil when the user is offline, or the Azure MetaData Service does not have this
+	// information and as such this should be used as best-effort, rather than guaranteed
+	Locations *[]string
+}
+
+type cloudEndpoint struct {
+	Endpoint  string    `json:"endpoint"`
+	Locations *[]string `json:"locations"`
+}
+
+type metaDataResponse struct {
+	CloudEndpoint map[string]cloudEndpoint `json:"cloudEndpoint"`
+}
+
+// AvailableAzureLocations returns a list of the Azure Locations which are available on the specified endpoint
+func AvailableAzureLocations(ctx context.Context, endpoint string) (*SupportedLocations, error) {
+	uri := fmt.Sprintf("https://%s//metadata/endpoints?api-version=2018-01-01", endpoint)
+	client := http.Client{
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+		},
+	}
+	req, err := http.NewRequestWithContext(ctx, "GET", uri, nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := client.Do(req)
```

Comment:
note: this intentionally doesn't use Azure/go-autorest since we don't need it here (it's an unauthorized endpoint, and logging this isn't super valuable)